atc/db/worker_resource_cache.go

```diff
@@ -37,6 +37,23 @@ func (workerResourceCache WorkerResourceCache) FindOrCreate(tx Tx, sourceWorkerB
 		return nil, false, err
 	}
 	if found {
+		// If the found worker resource cache's worker base resource type id is
+		// 0, then means the original source has been invalidated, so we need to
+		// reset the worker resource cache with the new base resource type id.
+		if uwrc.WorkerBaseResourceTypeID == 0 {
+			_, err := psql.Update("worker_resource_caches").
```

Comment:
I don't quite get your point. In line 45, it updates `worker_base_resource_type_id`. With the example you used, when a pruned worker comes back with an upgraded base resource type, new worker_base_resource_type will be updated to the tuple.

Reply:
Sorry maybe I was confusing with my wording. In my head the situation is a worker has existing volumes that are tied to a `worker_resource_cache_id` and then if we update the worker with a new base resource type, when it comes back we want to invalidate the volumes that are tied any worker resource caches that are using the old version of the worker base resource type. If we just update the row for a `worker_resource_cache_id` when it gets a new version of the base resource type, then it won't invalidate those caches and gc the volumes. I was thinking maybe we need to delete the old `worker_resource_cache_id` row entirely and create a new one so that it will gc the volumes that are pointed to the `worker_resource_cache_id` using the old base resource type version. Let me know if that makes sense or not!

Reply:
@clarafu I get you now.  I think the case you described is though valid, but should be almost impossible to happen.

* I believe in 99.9% of cases, base resource type's version doesn't matter to a resource cache.
* When a worker is pruned, will it come back with the same name and with upgraded base resource type? If a deploy needs to upgrade base resource type, it may just land workers but prune workers. In our deployment, a pruned worker's name will never been reused.

But I agree this is a potential issue, we should leave it there. I thought over, and am proposing 3 solutions:

1. A relative complicated solution. Currently, there is an unique index `(worker_name, resource_cache_id)` on table `worker_resource_cache`. As base resource type version matters, we should add base resource type id to this index. So that, we need to add a new column, say `worker_base_resource_type_id_orig`. When a worker_resource_cache tuple is inserted, `worker_base_resource_type_id_orig` holds the same value as `worker_base_resource_type_id`. When a worker is pruned, `worker_base_resource_type_id` will be set to null, but `worker_base_resource_type_id_orig` will keep the value unchanged. If a worker comes back, `worker_base_resource_type_id` should must change, so that it should create a new tuple in `worker_resource_cache`.
2. The second solution is relative simple. Given the case should be very rare, we can just delete the old tuple from `worker_resource_caches`, then insert a new tuple. Cons is that, as related volumes will be deleted on cascade, which may lead to some build failure. Pros is that, first simple, second safe. As the worker may come back with different team_id, tag, etc., and deleting old tuple will eliminate all potential problems.
3. Keep my current code and suggest users not reuse worker name.

WDYT? Or if you have other solution, we can discuss.

Reply:
For the second solution, if there is a way to tell the reason of the build failure i.e. user knows it is dued to pruning worker, then I would prefer this one. 

Third one will put more maintenance effort to OP and reusing worker name is very common e.g. all concourse.ci workers that deployed by k8s are with dedicate names.

Reply:
@xtremerui I don't see a way. Because when a `worker_resource_caches` tuple is deleted, some `volumes` tuples foreign-key pointing to the tuple will be deleted on cascade. Then build failure will not something like "volume not found".

If reusing worker name is common, is pruning a worker and re-adding it without upgrading anything common? If worker is only pruned for upgrading, then some build failures during upgrading can be considered as expected.

Otherwise we will have to go with solution 1.

Reply:
@xtremerui @clarafu Today I got some time, so I thought over this problem and realized that what are discussing is actually not a problem. Let me use an example to describe:

Say, worker-1 has initialized a worker resource cache:

| worker_base_resource_type_id|resource_cache_id|worker_name|
|--------------------------------|------------------|--------------|
| 1 | 1 | worker-1|

Then the cache is streamed to worker-2 and worker-3:

| worker_base_resource_type_id|resource_cache_id|worker_name|
|--------------------------------|------------------|--------------|
| 1 | 1 | worker-1|
| 1 | 1 | worker-2|
| 1 | 1 | worker-3|

Without this PR, when worker-1 is pruned, all 3 tuples will be deleted on cascade, which caused the bug.

With this PR, the first tuple will be deleted due to the foreign key `worker-name`, the other 2 tuples' worker-base-resource-type-id will be set to null:

| worker_base_resource_type_id|resource_cache_id|worker_name|
|--------------------------------|------------------|--------------|
| null | 1 | worker-2|
| null | 1 | worker-3|

So @clarafu 's concern about pruned worker should not be a problem.

Then the problem is, will worker-2 or worker-3 re-create the cache? The answer should be No. `workerResourceCache.FindOrCreate()` is only called in `volume.initializeResourceCache()`. In this PR, I have added a logic `createdVolume.InitializeStreamedResourceCache()`, when a found cache's worker-resource-type-id is null, the cache will not be initialized again.

So, here, when `uwrc.WorkerBaseResourceTypeID == 0`, instead of updating it, we can just raise an error.

Reply:
Right, I think the part for not initializing resource caches that have an invalid `worker_base_resource_type_id` works perfectly. I think the part I am concerned about is the usage of volumes that are tied to a `worker_resource_cache` that has an invalid or older `worker_base_resource_type_id` in the get step. 

For example, in the get step we will attempt to fetch from the cache

https://github.com/concourse/concourse/blob/a01d87ad47e916dd592a50d39cc1dd3c8ba70145/atc/exec/get_step.go#L327-L330

And the way we try to fetch from the cache is by looking to see if the volume exists on the worker

https://github.com/concourse/concourse/blob/a01d87ad47e916dd592a50d39cc1dd3c8ba70145/atc/exec/get_step.go#L406-L410

And then it will find a volume for that resource cache id on the worker

https://github.com/concourse/concourse/blob/a01d87ad47e916dd592a50d39cc1dd3c8ba70145/atc/db/volume_repository.go#L457-L462

And in your code, the `Find` method will return back a worker resource cache that has a 0 value for the `worker_base_resource_type_id`. This will then return the volume back and end up using this volume as the cache for the get step. 

Let me know if I got any of this wrong or I misinterpreted something!

Reply:
@clarafu First of all, this is a good discussion. I think you are wrong at step 2.

When `EnableCacheStreamedVolumes` is on, it `get` step will not select a worker:

https://github.com/concourse/concourse/blob/a01d87ad47e916dd592a50d39cc1dd3c8ba70145/atc/exec/get_step.go#L283

So then reaching to 

https://github.com/concourse/concourse/blob/a01d87ad47e916dd592a50d39cc1dd3c8ba70145/atc/exec/get_step.go#L406-L410

as `worker == nil`, it will go with line 407: `step.workerPool.FindResourceCacheVolume()`.

Then `step.workerPool.FindResourceCacheVolume()` will call `team.FindWorkersForResourceCache(resourceCache.ID())`. In this PR, I have modified this function by adding `sq.NotEq{"wrc.worker_base_resource_type_id": nil},`, so that if a worker cache's based resource type id is null, it won't be selected, so that `get` will not it as cache.

The idea of this PR is, when a worker resource cache has been used by a `get` step after source worker is pruned, then the cache can still be used in order to not break the running build; but the cache will no longer be used by new `get` as cache.

Reply:
@evanchaoli Oh yes thanks for explaining that!! 😊 Sorry I did get it wrong, I went to the wrong method. Rather than going to `workerPool.FindResourceCacheVolume()` I went to `volumeRepository.FindResourceCacheVolume()`.